File: elements-of-programming-interviews/cpp/reverse_sublist.cc
#include "list_node.h"
#include "test_framework/generic_test.h"

/*
0    1    2    3    4    5    6
A -> B -> C -> D -> E -> F -> G

ReverseSublist(2, 5)

0    1    2    3    4    5    6
A -> B -> F -> E -> D -> C -> G



- Have two pointers, to 5 (end, final) and 2 (start).
while (start != final)
- Set 2->next to 5->next
  - (start->next = end->next; end = start, start = start->next)
- Then set 3->next to 2
- Then set 4->next to 3
- Then set 5->next to 4


*/

shared_ptr<ListNode<int>> ReverseSublist(shared_ptr<ListNode<int>> L, int start,
                                         int finish) {
  // TODO - you fill in here.
  return L;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"L", "start", "finish"};
  return GenericTestMain(args, "reverse_sublist.cc", "reverse_sublist.tsv",
                         &ReverseSublist, DefaultComparator{}, param_names);
}

File: elements-of-programming-interviews/python/reverse_sublist.py
from typing import Optional

from list_node import ListNode
from test_framework import generic_test


def reverse_list(start: ListNode, end: ListNode) -> None:
    if start == end:
        return
    prev = None
    curr = start
    while prev != end:
        nextNext = curr.next
        curr.next = prev
        prev = curr
        curr = nextNext


def reverse_sublist(L: ListNode, start: int,
                    finish: int) -> Optional[ListNode]:
    if not start or start == finish:
        return L
    pre_start_node = start_node = end_node = post_end_node = None
    curr = L
    i = 1
    while curr:
        if i == start-1:
            pre_start_node = curr
        elif i == start:
            start_node = curr
        elif i == finish:
            end_node = curr
        elif i == finish+1:
            post_end_node = curr
        i += 1
        curr = curr.next

    reverse_list(start_node, end_node)
    # if there was a pre-start node, it points now to end node
    if pre_start_node:
        pre_start_node.next = end_node
    # if there was a post-end node, the old start node now points to it.
    if post_end_node:
        start_node.next = post_end_node
    return L if pre_start_node else end_node


if __name__ == '__main__':
    """
    start_i = 1
    finish_i = 9
    head = start = finish = end = ListNode(0, None)

    for i in range(1, 10):
        end.next = ListNode(i, None)
        end = end.next
        if i == start_i:
            start = end
        if i == finish_i:
            finish = end
    reverse_sublist(head, start_i, finish_i)

    while (head):
        print(f"Node: {head.data}")
        head = head.next

    """
    exit(
        generic_test.generic_test_main('reverse_sublist.py',
                                       'reverse_sublist.tsv', reverse_sublist))
